src/ucp/api/ucp_def.h

```diff
@@ -81,6 +82,29 @@ typedef struct ucp_config                ucp_config_t;
 typedef struct ucp_ep                    *ucp_ep_h;
 
 
+/**
+ * @ingroup UCP_ENDPOINT
+ * @brief Callback to process incoming active message
+ *
+ * @param [in]  arg     User-defined argument.
+ * @param [in]  data    Points to the received data. 
+ * @param [in]  length  Length of data.
+ * @param [in]  flags   Ignore, these are only relevant at the UCT level
+ *
+ * @note This callback could be set and released
+ *       by @ref ucp_worker_set_am_handler function.
+ *       The minimum value of length is 8. If a user only
+ *       specifies to send 1 byte, that will be the only
+ *       valid byte in data, however length will still be set to 8.
+ *
+ * @retval UCS_OK         - The user must always specify the function to
+ *                          return UCS_OK. Failure to do this can result
+ *                          memory leaks or errors
+ */
+typedef ucs_status_t (*ucp_am_callback_t)(void *arg, void *data, size_t length,
```

Comment:
seems like the callback is called when all data is received, this may be a problem since then UCP would have to allocate a contiguous buffer and copy the data into it.
how about getting iovec here, or adding another callback to process fragments of long data like in PAMI?

Reply:
@yosefe can you please dump PAMI example here.

Reply:
i don't have something which is publicly available

Reply:
why not just use stream API?
**pros:** 2 ways to handle data - `ucp_stream_recv_data_nb` for fragmented recvs and `ucp_stream_recv_nb` + wait_all for message like
**cons:** endpoints should be connected pairwise + extra header is needed

or implement something similar without callbacks from UCT?

Reply:
we are wanting to use an AM api for one-sided program models and internal (internal to things like Open SHMEM) control messages, kind of like what's in GasNET. stream api does not match well with this type of use case.

Reply:
@yosefe I like the idea of a callback for long messages. Do you want to propose any specific API ? Some example will help.

Reply:
@yosefe found this https://www.ibm.com/support/knowledgecenter/en/SSFK3V_2.3.0/com.ibm.cluster.protocols.v2r3.pp400.doc/bl510_pdisps.htm

But i guess you are talking about something else